client/app/services/parameters/Parameter.js

```diff
@@ -0,0 +1,77 @@
+import { isNull, isObject, isFunction, has } from 'lodash';
+import { TextParameter, NumberParameter } from '.';
+
+class Parameter {
+  constructor(parameter, parentQueryId) {
+    this.title = parameter.title;
+    this.name = parameter.name;
+    this.type = parameter.type;
+    this.global = parameter.global; // backward compatibility in Widget service
+    this.parentQueryId = parentQueryId;
+
+    // Used for meta-parameters (i.e. dashboard-level params)
+    this.locals = [];
+
+    // Used for URL serialization
+    Object.defineProperty(this, 'urlPrefix', {
+      configurable: true,
+      enumerable: false, // don't save it
+      writable: true,
+      value: 'p_',
+    });
+  }
+
+  static create(param) {
+    switch (param.type) {
+      case 'text':
+        return new TextParameter(param);
+      case 'number':
+        return new NumberParameter(param);
+      default:
+        return null;
+    }
+  }
```

Comment:
This seems like an anti-pattern, no?
The super class should not be aware of its heirs.

Reply:
:thinking: seemed ok as it was `static`, the goal is to have a factory method, can move it to an external function in any case :)

Reply:
It's a matter of taste :) Whatever you see fit 👍